pkg/lib/authn/authenticator/whatsapp/provider.go

```diff
@@ -0,0 +1,85 @@
+package whatsapp
+
+import (
+	"errors"
+
+	"github.com/authgear/authgear-server/pkg/util/clock"
+	"github.com/authgear/authgear-server/pkg/util/log"
+	"github.com/authgear/authgear-server/pkg/util/secretcode"
+)
+
+type Logger struct{ *log.Logger }
+
+func NewLogger(lf *log.Factory) Logger { return Logger{lf.New("whatsapp")} }
+
+type Provider struct {
+	CodeStore *StoreRedis
+	Clock     clock.Clock
+	Logger    Logger
+}
+
+func (p *Provider) GetCode(phone string) (*Code, error) {
+	return p.CodeStore.Get(phone)
+}
+
+func (p *Provider) CreateCode(phone string, appID string, webSessionID string) (*Code, error) {
+	code := secretcode.OOBOTPSecretCode.Generate()
+	codeModel := &Code{
+		AppID:        appID,
+		WebSessionID: webSessionID,
+		Phone:        phone,
+		Code:         code,
+		ExpireAt:     p.Clock.NowUTC().Add(WhatsappCodeDuration),
+	}
+
+	err := p.CodeStore.Create(codeModel)
+	if err != nil {
+		return nil, err
+	}
+	return codeModel, nil
+}
+
+func (p *Provider) VerifyCode(phone string, webSessionID string) (*Code, error) {
+	codeModel, err := p.CodeStore.Get(phone)
+	if errors.Is(err, ErrCodeNotFound) {
+		return nil, ErrInvalidCode
+	} else if err != nil {
+		return nil, err
+	}
+
+	if webSessionID != codeModel.WebSessionID {
+		return nil, ErrWebSessionIDMismatch
+	}
+
+	if codeModel.UserInputtedCode == "" {
+		return nil, ErrInputRequired
+	}
+
+	if !secretcode.OOBOTPSecretCode.Compare(codeModel.UserInputtedCode, codeModel.Code) {
+		return nil, ErrInvalidCode
+	}
+
+	return codeModel, nil
+}
+
+func (p *Provider) DeleteCode(phone string) {
```

Comment:
This method should not be exposed. Instead, in VerifyCode, if the code is correct, delete code from the store directly. This can ensure code is always one-time use

Reply:
Since we will need to verify the code without consuming for switching the state in the UI, so I separate it. Let me change the verify function to `VerifyCode(phone string, webSessionID string, consume bool)`, so we don't have to expose the delete function